File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_word = False
        

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for ch in word:
            node = node.children.setdefault(ch, TrieNode())
        node.is_word = True

    def search(self, word: str) -> bool:
        return self.dfs(0, word, self.root)
      
    def dfs(self, index, word, node):
        if index == len(word):
            return node.is_word
        ch = word[index]
        if ch == '.':
            for child in node.children:
                if self.dfs(index + 1, word, node.children[child]):
                    return True
        if ch in node.children:
            return self.dfs(index + 1, word, node.children[ch])
        return False


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.js
class TrieNode {
  constructor() {
    this.children = {};
    this.isWord = false;
  }
}


var WordDictionary = function() {
  this.root = new TrieNode();
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {
  let node = this.root;
  
  for (let ch of word) {
    if (!node.children[ch]) node.children[ch] = new TrieNode();
    
    node = node.children[ch];
  }
  
  node.isWord = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
  const dfs = (index, node) => {
    if (index === word.length) return node.isWord;
    
    const ch = word[index++];
    if (ch === '.') 
      for (let child of Object.values(node.children)) 
        if (dfs(index, child)) return true;
    if (node.children[ch]) return dfs(index, node.children[ch]);
    
    return false;
  }
  
  return dfs(0, this.root);
};

/** 
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */